### PR TITLE
IGNITE-17604 C++ Client transactions

### PR DESCRIPTION
**Pull request by isapego**

### CODE REVIEW

File: modules/platforms/cpp/ignite/client/detail/transaction/transaction_impl.h [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        state old = state::OPEN;
+        auto opened = m_state.compare_exchange_strong(old, st, std::memory_order_relaxed);
+        if (opened)
+            return;
+
+        auto message = old == state::COMMITTED
+            ? "Transaction is already committed."
+            : "Transaction is already rolled back.";
+
+        throw ignite_error(message);
```
Comment (ptupitsyn) [inline modules/platforms/cpp/ignite/client/detail/transaction/transaction_impl.h:139]: This behavior will be changed soon on the Java side: https://issues.apache.org/jira/browse/IGNITE-18324
As I understand, commit/rollback will be ignored if a tx is already committed or rolled back.

File: modules/platforms/cpp/tests/client-test/main.cpp [hunk 2]
```diff
@@ -74,21 +74,21 @@ void before_all() {
 
 int main(int argc, char **argv) {
     int res = 0;
-    before_all();
+//    before_all();
```
Comment (ptupitsyn) [inline modules/platforms/cpp/tests/client-test/main.cpp:77]: Is this a temporary change?

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/PlatformTestNodeRunner.java [hunk 3]
```diff
@@ -73,12 +73,12 @@ public class PlatformTestNodeRunner {
     private static final String TABLE_NAME_ALL_COLUMNS_SQL = "tbl_all_columns_sql"; // All column types supported by SQL.
 
     /** Time to keep the node alive. */
-    private static final int RUN_TIME_MINUTES = 30;
+    private static final int RUN_TIME_MINUTES = 300;
```
Comment (ptupitsyn) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/PlatformTestNodeRunner.java:76]: The idea behind 30 minutes is to deal with hanging processes on TeamCity - the process will exit by itself after some time. Not sure if 300 minutes makes any sense.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/PlatformTestNodeRunner.java [hunk 4]
```diff
@@ -73,12 +73,12 @@ public class PlatformTestNodeRunner {
...
 
     /** Nodes bootstrap configuration. */
     private static final Map<String, String> nodesBootstrapCfg = Map.of(
             NODE_NAME, "{\n"
-                    + "  \"clientConnector\":{\"port\": 10942,\"portRange\":10,\"idleTimeout\":3000,\""
+                    + "  \"clientConnector\":{\"port\": 10942,\"portRange\":10,\"idleTimeout\":0,\""
```
Comment (ptupitsyn) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/PlatformTestNodeRunner.java:81]: Just curious - what's the reason behind this change?

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

File: modules/platforms/cpp/ignite/client/detail/transaction/transaction_impl.h (hunk 1 above)
Comment (isapego) [reply modules/platforms/cpp/ignite/client/detail/transaction/transaction_impl.h:139]: Looks weird, but OK, let's change it too.

File: modules/platforms/cpp/tests/client-test/main.cpp (hunk 2 above)
Comment (isapego) [reply modules/platforms/cpp/tests/client-test/main.cpp:77]: Damn, yeah, it was for the debugging. I'll reverse it.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/PlatformTestNodeRunner.java (hunk 3 above)
Comment (isapego) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/PlatformTestNodeRunner.java:76]: It was for the debugging as well :) Reversing it.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/PlatformTestNodeRunner.java (hunk 4 above)
Comment (isapego) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/PlatformTestNodeRunner.java:81]: The same as above -- debugging :)

File: modules/platforms/cpp/ignite/client/detail/transaction/transaction_impl.h (hunk 1 above)
Comment (ptupitsyn) [reply modules/platforms/cpp/ignite/client/detail/transaction/transaction_impl.h:139]: We can probably do it later after Java-side changes are finalized and merged, to ensure consistent behavior.

**Review by ptupitsyn — APPROVED** [review]
Ship it :+1:
